snapcraft/internal/errors.py

```diff
@@ -31,6 +31,56 @@
 CLEAN_RESOLUTION = "Run `snapcraft clean` and retry build."
 
 
+def details_from_command_error(
+    *,
+    cmd: List[str],
+    returncode: int,
+    stdout: Optional[Union[bytes, str]] = None,
+    stderr: Optional[Union[bytes, str]] = None,
+) -> str:
+    """Create a consistent details message from command errors.
+
+    stdout and stderr, if provided, will be stringified using its object
+    representation.  This method does not decode byte strings.
+
+    :param cmd: Command executed.
+    :param returncode: Command exit code.
+    :param stdout: Optional stdout to include.
+    :param stderr: Optional stderr to include.
+
+    :returns: Details string.
+    """
+    cmd_string = " ".join([shlex.quote(c) for c in cmd])
+
+    details = [
+        f"* Command that failed: {cmd_string!r}",
+        f"* Command exit code: {returncode}",
+    ]
+
+    if stdout:
+        details.append(f"* Command output: {stdout!r}")
+
+    if stderr:
+        details.append(f"* Command standard error output: {stderr!r}")
+
+    return "\n".join(details)
```

Comment:
This is suspiciously unbalanced: all lines end with a `\n` except the last one. 

How this will be used? It's ok for last line to not have an ENTER?

Reply:
Correct, none of the exception fields have trailing newlines in Snapcraft.  The top level handler will print with the newline.